vsphere/resource_vsphere_vm_storage_policy.go

```diff
@@ -0,0 +1,300 @@
+		if strings.Contains(err.Error(), "Profile not found") {
+			log.Printf("[DEBUG] storage policy profile %s: Resource has been deleted", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("error while reading vm storage policy profile with Id %s %s", d.Id(), err.Error())
+	}
+	if len(vmStoragePolicies) == 0 {
+		d.SetId("")
+		return nil
+	}
+
+	pbmCapabilityProfile := vmStoragePolicies[0].(*types2.PbmCapabilityProfile)
+	d.SetId(pbmCapabilityProfile.ProfileId.UniqueId)
+	d.Set("name", pbmCapabilityProfile.Name)
+	d.Set("description", pbmCapabilityProfile.Description)
+
+	var tagRules []map[string]interface{}
+	if pbmCapabilityProfile.Constraints != nil {
+		pbmSubProfileConstraints := pbmCapabilityProfile.Constraints.(*types2.PbmCapabilitySubProfileConstraints)
+		if pbmSubProfileConstraints == nil {
+			return nil
+		}
+		pbmSubProfiles := pbmSubProfileConstraints.SubProfiles
+		for _, subProfile := range pbmSubProfiles {
+			if subProfile.Name != TAG_PLACEMENT {
+				continue
+			}
+			capabilities := subProfile.Capability
+			for _, capability := range capabilities {
+
+				if capability.Id.Namespace != TAG_NAMESPACE {
+					continue
+				}
+				tagCategory := capability.Id.Id
+
+				constraints := capability.Constraint
+				if len(constraints) == 0 {
+					continue
+				}
+				propertyInstances := constraints[0].PropertyInstance
+				if len(propertyInstances) == 0 {
+					continue
+				}
+				tagsSet := propertyInstances[0].Value.(types2.PbmCapabilityDiscreteSet)
+				tagRule := make(map[string]interface{})
+				tagRule["tag_category"] = tagCategory
+				tagRule["tags"] = tagsSet.Values
+
+				includeTags := true
+				if propertyInstances[0].Operator == "NOT" {
+					includeTags = false
+				}
+				tagRule["include_datastores_with_tags"] = includeTags
+
+				tagRules = append(tagRules, tagRule)
+			}
+		}
+	}
+	d.Set("tag_rules", tagRules)
+	return nil
+}
+
+func resourceVmStoragePolicyUpdate(d *schema.ResourceData, meta interface{}) error {
+	log.Print("[DEBUG] :  Performing update")
+	client := meta.(*VSphereClient).vimClient
+	rc := meta.(*VSphereClient).restClient
+	pbmClient, err := pbm.NewClient(context.Background(), client.Client)
+	if err != nil {
+		return fmt.Errorf("error while creating pbm client %s", err)
+	}
+
+	updateSpec := types2.PbmCapabilityProfileUpdateSpec{}
+
+	updateSpec.Name = d.Get("name").(string)
+	updateSpec.Description = d.Get("description").(string)
+	log.Print("update spec ", updateSpec)
+
+	if d.HasChange("tag_rules") {
+
+		tagsManager := tags.NewManager(rc)
+		tagRules := d.Get("tag_rules").([]interface{})
+		var capabilities []pbm.Capability
+
+		for _, tagRule := range tagRules {
+
+			tagCategory := tagRule.(map[string]interface{})["tag_category"].(string)
+			_, err := tagCategoryByName(tagsManager, tagCategory)
+			if err != nil {
+				return fmt.Errorf("error while getting the tag %s %s", tagCategory, err)
+			}
+
+			tagValues := tagRule.(map[string]interface{})["tags"].([]interface{})
```

Comment:
what happens when required params tag_category and tags are not provided?

Reply:
tag_category is required, so terraform plan will throw an error.
Same with tags. Minimum 1 element is required.